lib/zhr_devs/identity_management/aggregates/identity.ex

```diff
@@ -0,0 +1,46 @@
+defmodule ZhrDevs.IdentityManagement.Aggregates.Identity do
+  import Algae
+
+  defdata do
+    identity :: Uptight.Text.t()
+    hashed_identity :: Uptight.Base.Urlsafe.t()
+    login_at :: UtcDateTime.t()
+  end
+
+  alias ZhrDevs.IdentityManagement.{Events, Commands}
+
+  alias Uptight.Result
+  alias Uptight.Text, as: T
+
+  alias ZhrDevs.IdentityManagement.Aggregates.Identity
+
+  def execute(%Identity{}, %Commands.Login{} = login_command) do
+    login_command
+    |> parse_success()
+    |> handle_result()
+  end
+
+  def apply(%Identity{} = identity, %Events.LoggedIn{} = logged_in) do
+    %Identity{identity | login_at: logged_in.login_at}
+  end
+
+  #### Private functions ####
+
+  defp parse_success(%Commands.Login{identity: identity, hashed_identity: hashed_identity}) do
+    Result.new(fn ->
+      identity = identity |> T.new() |> Result.from_ok()
+      hashed_identity = hashed_identity |> Uptight.Base.mk_url() |> Result.from_ok()
+
+      %Events.LoggedIn{
+        identity: identity,
+        hashed_identity: hashed_identity,
+        login_at: UtcDateTime.new()
+      }
+    end)
+  end
+
+  @parse_err "Parsing OAuth success struct failed"
+
+  defp handle_result(%Result.Ok{} = ok), do: Result.from_ok(ok)
```

Comment:
I probably should put this in Uptight.Result, right? Like a function `get_ok(error_override \\ nil)`.